README.md
# on-air

Show a warning light when someone in the room is **on-air**, using [Google Cloud Pub/Sub](https://cloud.google.com/pubsub) and [blink(1)](https://blink1.thingm.com/).

## Getting started

### Google Cloud project

You will need to create a [Google Cloud Pub/Sub](https://cloud.google.com/pubsub) resource to act as a message queue.

You will also need to provision a service accounts for each device in the system:

- one or more publisher credentials, to write events
- one subscriber credential, to read events

This is currently outside the scope of this readme.

### Local clients

The clients are both contained in a single python script. Some third party dependencies are required.

```bash
# Clone this repo
git clone https://github.com/tommilligan/on-air
cd on-air

# Instll third-party deps
pip install -r requirements.txt
```

On devices with hardware to monitor, start streaming events up to the cloud:

```bash
# The variables you will need to fill in from your GCP project
GOOGLE_PROJECT_ID=google-project-012345
GOOGLE_CREDENTIAL=publisher-credential-file.json
TOPIC_NAME=topic-name

# This will run forever, watching for changes. Use Ctrl+C to exit.
./on_air.py  \
  --google-project-id "$GOOGLE_PROJECT_ID"  \
  --google-credential "$GOOGLE_CREDENTIAL"  \
  stream  \
  --poll-interval 5  \
  --topic-name "$TOPIC_NAME"
```

On a device with a [blink(1)](https://blink1.thingm.com/) indicator plugged in, start listening for events:

```bash
# The variables you will need to fill in from your GCP project
GOOGLE_PROJECT_ID=google-project-012345
GOOGLE_CREDENTIAL=subscriber-credential-file.json
SUBSCRIPTION_NAME=subscription-name

# This will run forever, watching for changes. Use Ctrl+C to exit.
./on_air.py  \
  --google-project-id "$GOOGLE_PROJECT_ID"  \
  --google-credential "$GOOGLE_CREDENTIAL"  \
  listen  \
  --subscription-name "$SUBSCRIPTION_NAME"
```

## blink1 setup

For a linux machine, you will need to add udev rules for the device. A copy is in this repo:

```bash
sudo cp 51-blink1.rules /etc/udev/rules.d/
sudo udevadm control --reload
```

## Raspberry Pi Further Setup

Some additional steps and dependencies are needed for Raspberry Pi:

```bash
sudo apt-get install  \
  python3-venv
cd ~
mkdir on-air
cd on-air
python3 -m venv env
```

### Systemd Services

Example service files are included for both listener (rpi) and streamer.

### Log rotation

You should set up log rotation like:

```conf
# /etc/logrotate.d/on-air-listen

/home/pi/on-air/on-air.log {
  daily
  missingok
  rotate 4
  compress
  copytruncate
}
```

## Architecture

The architecture used is a multiple-publisher single-consumer pub/sub flow:

```
                 +----------------+
                 |                |
                 |  PubSub Queue  |
                 |                |
                 +--^----------+--+
                    |          |
                    |          |
         +----------+--+    +--v-----------+
         |             |    |              |
         |  Publisher  |    |  Subscriber  |
         |             |    |              |
         +----------^--+    +--+-----------+
                    |          |
                    |          |
+-------------------+-+      +-v---------------+
|                     |      |                 |
|  Camera/Microphone  |      |  LED Indicator  |
|                     |      |                 |
+---------------------+      +-----------------+
```

Changes to hardware usage on the publisher device are polled for using [`lsof`](https://man7.org/linux/man-pages/man8/lsof.8.html) to watch the devices mounted in [`/dev`](https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/dev.html).
When a change is detected, a message is published to a topic in [Google Cloud Pub/Sub](https://cloud.google.com/pubsub).

A subscriber interested in changes listens for messages on a linked topic using a [`google` gRPC streaming client](https://github.com/googleapis/python-pubsub).
On recieving a message, the subscriber caches messages by source. It then computes an additive state for all publishers, and triggers a user notification via the [blink(1)](https://blink1.thingm.com/) indicator if applicable.

## Requirements

- Low latency: less than 10 seconds latency from change to notification
  - The limiting factor is how fast the publisher polls for hardware changes
  - If a very short poll time is chosen, there may be some interesting effects as LED updates overlap
- Low throughput: stay under the GCP free tier
  - To reduce messaging volume, the publisher caches local state and only publishes a new message on change
  - Data payload size could be reduced by encoding the payload using something like protobuf (currently JSON)
- Fail open: if state is unknown, default to indicator off
  - On subscriber crash or restart, the indicator is reset

## Configuration

For a streamer the easiest installation is to install the debian package. This will automatically install a service daemon to stream events in the background.

Additional configuration is required for this method:

- Create a system directory for configuration: `sudo mkdir /etc/on-air`
- Move your GCP JWT credential here: `sudo mv google-credential.json /etc/on-air/google-credential.json`
- Set up a config file at `/etc/on-air/on-air.json`. Example contents:

```json
{
  "google_project_id": "google-project-012345",
  "google_credential": "/etc/on-air/google-credential.json",
  "poll_interval": 5,
  "topic_name": "topic-name"
}
```

## TODO List

- [x] Support multiple publisher to single subscriber flow
- [x] Servicify clients using something like a `systemd` service
- [x] One click/command install of python clients
- [ ] Better instructions/automated setup of GCP resources
  - [ ] Subscribers create their own subscription topic
- [ ] Configurable notification settings (color/blink count)
  - [x] Load settings from config file


debian/changelog
on-air (0.1.1-1) UNRELEASED; urgency=medium

  * Initial release. (Closes: #XXXXXX)
  * Add missing dependency on python3-distutils
  * Fix bug where audio playback devices were also monitored

 -- Tom <tom@calvados>  Sun, 21 Feb 2021 19:16:28 +0000


on_air.py
#!/usr/bin/env python

import argparse
import glob
import json
import logging
import signal
import socket
import subprocess
import sys
import time
from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any, Callable, DefaultDict, Dict, Iterable, List, Optional, Tuple

import re
from blink1.blink1 import Blink1
from google.auth import jwt
from google.cloud import pubsub

Payload = Dict[str, Any]

_log = logging.getLogger(__file__)

_LSOF_KNOWN_ERRORS = re.compile(
    r"(?:Output information may be incomplete\.|WARNING: can't stat\(\) .* file system)"
)


class OnAirError(Exception):
    pass


class SystemError(OnAirError):
    pass


def lsof(pattern: str) -> List[str]:
    """List file owner PIDs for the given glob pattern."""
    files = glob.glob(pattern)
    response = subprocess.run(
        ["/usr/bin/lsof", "-t"] + files,
        capture_output=True,
        text=True,
    )
    if response.returncode != 0:
        # Expect some errors, these are fine
        for line in response.stderr.split("\n"):
            if len(line.strip()) == 0:
                continue

            if not _LSOF_KNOWN_ERRORS.search(line):
                _log.error(line)
                raise SystemError(
                    f"Unexpected error from lsof: '{line}', "
                    f"stdout '{response.stdout}', stderr '{response.stderr}'"
                )

    stdout = response.stdout.rstrip()
    if len(stdout) > 0:
        owners = stdout.split("\n")
        return owners

    return []


def poll_av_and_publish(
    poll_interval: int,
    publish_payload: Callable[[Payload], None],
    source_name: str,
) -> None:
    """Poll local audio/video hardware in a loop, and publish update messages."""

    # This will loop forever, so set a nice shutdown handler
    signal.signal(signal.SIGINT, shutdown)

    last_payload: Optional[Payload] = None

    _log.info("Watching for changes in local audio/video state")
    while True:
        _log.debug("Polling local audio/video state")
        audio_owners = lsof("/dev/snd/pcmC*")
        video_owners = lsof("/dev/video*")

        payload = {
            "audio": len(audio_owners) > 0,
            "video": len(video_owners) > 0,
            "source": source_name,
        }

        if payload != last_payload:
            last_payload = payload
            publish_payload(payload)

        time.sleep(poll_interval)


def run_stream(args: argparse.Namespace) -> None:
    """Entrypoint for the streaming client."""
    service_account_info = json.load(open(args.google_credential))
    audience = "https://pubsub.googleapis.com/google.pubsub.v1.Publisher"
    credentials = jwt.Credentials.from_service_account_info(
        service_account_info, audience=audience
    )
    publisher = pubsub.PublisherClient(credentials=credentials)
    topic_name = f"projects/{args.google_project_id}/topics/{args.topic_name}"

    def publish_payload(payload: Payload) -> None:
        data = json.dumps(payload)
        _log.info("Publishing message: '%s'", data)
        future = publisher.publish(topic_name, data.encode("utf-8"))
        future.result()

    poll_av_and_publish(
        poll_interval=args.poll_interval,
        publish_payload=publish_payload,
        source_name=args.source_name,
    )


# Time for an interval of blinking, seconds
_BLINK_DURATION = 0.1
# Number of alerting blinks before solid color
_BLINK_REPEAT = 3

Rgb = Tuple[int, int, int]
_RGB_VIDEO = (255, 0, 0)
_RGB_AUDIO = (0, 0, 255)
_RGB_CLEAR = (0, 255, 0)
_RGB_OFF = (0, 0, 0)


_HARDWARE_KEYS = ("audio", "video")


@dataclass(eq=True, frozen=True)
class ComputedState:
    audio: bool
    video: bool

    @staticmethod
    def from_source_states(source_states: Iterable[Payload]) -> "ComputedState":
        audio = False
        video = False

        for state in source_states:
            if state["audio"]:
                audio = True
            if state["video"]:
                video = True

        return ComputedState(audio=audio, video=video)


class DisplayState:
    _device: Optional[Blink1]
    # Map of {source name -> Payload}
    _source_states: DefaultDict[str, Payload]
    # Computed hash over all source states
    _state: ComputedState

    def __init__(self, device):
        self._device = device
        self._source_states = defaultdict(dict)
        self._state = ComputedState(audio=False, video=False)

    def __enter__(self, *args):
        return self

    def __exit__(self, *args):
        if self._device:
            self._device.off()

    def _solid(self, color: Rgb) -> None:
        _log.debug("Color set: '%s'", color)
        if self._device:
            self._device.fade_to_rgb(0, *color)

    def _blink(self, color: Rgb) -> None:
        for _ in range(0, _BLINK_REPEAT):
            self._solid(color)
            time.sleep(_BLINK_DURATION)
            self._solid(_RGB_OFF)
            time.sleep(_BLINK_DURATION)

    def update(self, payload: Payload) -> None:
        source_name = payload["source"]
        previous_source_state = self._source_states[source_name]
        if previous_source_state == payload:
            _log.debug("Source state is unchanged: '%s'", payload)
            return
        self._source_states[source_name] = payload

        state = ComputedState.from_source_states(self._source_states.values())
        if self._state == state:
            _log.debug("Computed state is unchanged: '%s'", state)
            return
        self._state = state
        _log.debug("Computed state updated: '%s'", state)

        if state.video:
            color = _RGB_VIDEO
        elif state.audio:
            color = _RGB_AUDIO
        else:
            # Once wer're clear, flash an all clear, then turn off
            color = _RGB_CLEAR
            self._blink(color)
            self._solid(_RGB_OFF)
            return

        self._blink(color)
        self._solid(color)


_LISTEN_SKEW = timedelta(minutes=1)


def run_listen(args: argparse.Namespace) -> None:
    """Entrypoint for the listening client."""
    service_account_info = json.load(open(args.google_credential))
    audience = "https://pubsub.googleapis.com/google.pubsub.v1.Subscriber"
    credentials = jwt.Credentials.from_service_account_info(
        service_account_info, audience=audience
    )
    subscriber = pubsub.SubscriberClient(credentials=credentials)
    subscription_name = (
        f"projects/{args.google_project_id}/subscriptions/{args.subscription_name}"
    )

    if args.no_blink1:
        device = None
    else:
        device = Blink1()
        device.off()

    with DisplayState(device) as display_state:

        def recieve_message(message) -> None:
            now = datetime.now(tz=timezone.utc)
            if message.publish_time < (now - _LISTEN_SKEW):
                _log.debug(
                    "Discarding skewed message: now '%s', message '%s'",
                    now,
                    message.publish_time,
                )
                message.ack()
                return

            payload = message.data.decode("utf-8")
            _log.info("Recieved message: %s", payload)
            data = json.loads(payload)
            display_state.update(data)

            message.ack()

        # This will loop forever, so set a nice shutdown handler
        signal.signal(signal.SIGINT, shutdown)

        _log.info("Listening for published updates")
        future = subscriber.subscribe(subscription_name, recieve_message)
        future.result()


def shutdown(signal_number, frame):
    """Shutdown handler for system signals."""
    _log.info(
        "Received signal '%s', shutting down",
        signal_number,
    )
    sys.exit(1)


def main() -> None:
    """Main command line entrypoint."""
    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    stream_handler = logging.StreamHandler()
    stream_handler.setFormatter(
        logging.Formatter("%(asctime)s|%(name)s|%(levelname)s|%(message)s")
    )
    root_logger.handlers = [stream_handler]

    parser = argparse.ArgumentParser(description="API Gateway Service")
    parser.add_argument(
        "--google-project-id",
        type=str,
        required=True,
        help="Google project id to use",
    )
    parser.add_argument(
        "--google-credential",
        type=str,
        required=True,
        help="Google credential JWT file",
    )

    subparsers = parser.add_subparsers(required=True, dest="subcommand")
    stream = subparsers.add_parser("stream")
    stream.set_defaults(execute=run_stream)
    stream.add_argument(
        "--poll-interval",
        type=int,
        required=True,
        help="Interval to check local state in seconds",
    )
    stream.add_argument(
        "--topic-name",
        type=str,
        required=True,
        help="Topic to publish messages to",
    )
    stream.add_argument(
        "--source-name",
        type=str,
        default=socket.gethostname(),
        help="Specify the name of this source of events. Defaults to system hostname.",
    )

    listen = subparsers.add_parser("listen")
    listen.set_defaults(execute=run_listen)
    listen.add_argument(
        "--subscription-name",
        type=str,
        required=True,
        help="Google subscription to recieve messages from",
    )
    listen.add_argument(
        "--no-blink1",
        action="store_true",
        help="Do not use blink(1) for user notifications",
    )

    args = parser.parse_args()
    args.execute(args)


if __name__ == "__main__":
    main()


on_air/__init__.py
#!/usr/bin/env python

import argparse
import glob
import json
import logging
import re
import signal
import socket
import subprocess
import sys
import time
from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any, Callable, DefaultDict, Dict, Iterable, List, Optional, Tuple

from blink1.blink1 import Blink1
from google.auth import jwt
from google.cloud import pubsub

Payload = Dict[str, Any]

_log = logging.getLogger(__file__)

_LSOF_KNOWN_ERRORS = re.compile(
    r"(?:Output information may be incomplete\.|WARNING: can't stat\(\) .* file system)"
)

_GLOB_AUDIO_DEVICES = "/dev/snd/pcmC*D*c"
_GLOB_VIDEO_DEVICES = "/dev/video*"


class OnAirError(Exception):
    pass


class SystemError(OnAirError):
    pass


def lsof(pattern: str) -> List[str]:
    """List file owner PIDs for the given glob pattern."""
    files = glob.glob(pattern)
    response = subprocess.run(
        ["/usr/bin/lsof", "-t"] + files,
        capture_output=True,
        text=True,
    )
    if response.returncode != 0:
        # Expect some errors, these are fine
        for line in response.stderr.split("\n"):
            if len(line.strip()) == 0:
                continue

            if not _LSOF_KNOWN_ERRORS.search(line):
                _log.error(line)
                raise SystemError(
                    f"Unexpected error from lsof: '{line}', "
                    f"stdout '{response.stdout}', stderr '{response.stderr}'"
                )

    stdout = response.stdout.rstrip()
    if len(stdout) > 0:
        owners = stdout.split("\n")
        return owners

    return []


def poll_av_and_publish(
    poll_interval: int,
    publish_payload: Callable[[Payload], None],
    source_name: str,
) -> None:
    """Poll local audio/video hardware in a loop, and publish update messages."""

    # This will loop forever, so set a nice shutdown handler
    signal.signal(signal.SIGINT, shutdown)

    last_payload: Optional[Payload] = None

    _log.info("Watching for changes in local audio/video state")
    while True:
        _log.debug("Polling local audio/video state")
        audio_owners = lsof(_GLOB_AUDIO_DEVICES)
        video_owners = lsof(_GLOB_VIDEO_DEVICES)

        payload = {
            "audio": len(audio_owners) > 0,
            "video": len(video_owners) > 0,
            "source": source_name,
        }

        if payload != last_payload:
            last_payload = payload
            publish_payload(payload)

        time.sleep(poll_interval)


@dataclass
class Config:
    google_credential: str
    google_project_id: str
    poll_interval: int
    topic_name: str
    source_name: str

    @staticmethod
    def from_args(args: argparse.Namespace) -> "Config":
        google_credential = None
        google_project_id = None
        poll_interval = None
        topic_name = None
        source_name = None

        if args.config:
            with open(args.config, "r") as config_file:
                raw_config_json = json.load(config_file)

            google_credential = raw_config_json.get("google_credential")
            google_project_id = raw_config_json.get("google_project_id")
            poll_interval = raw_config_json.get("poll_interval")
            topic_name = raw_config_json.get("topic_name")
            source_name = raw_config_json.get("source_name")

        if args.google_credential:
            google_credential = args.google_credential
        if args.google_project_id:
            google_project_id = args.google_project_id
        if args.poll_interval:
            poll_interval = args.poll_interval
        if args.topic_name:
            topic_name = args.topic_name
        if args.source_name:
            source_name = args.source_name

        if not source_name:
            source_name = socket.gethostname()

        if google_credential is None or not isinstance(google_credential, str):
            raise ValueError(f"Invalid google_credential: '{google_credential}'")
        if google_project_id is None or not isinstance(google_project_id, str):
            raise ValueError(f"Invalid google_project_id: '{google_project_id}'")
        if poll_interval is None or not isinstance(poll_interval, int):
            raise ValueError(f"Invalid poll_interval: '{poll_interval}'")
        if topic_name is None or not isinstance(topic_name, str):
            raise ValueError(f"Invalid topic_name: '{topic_name}'")
        if source_name is None or not isinstance(source_name, str):
            raise ValueError(f"Invalid source_name: '{source_name}'")

        return Config(
            google_credential=google_credential,
            google_project_id=google_project_id,
            poll_interval=poll_interval,
            topic_name=topic_name,
            source_name=source_name,
        )


def run_stream(args: argparse.Namespace) -> None:
    """Entrypoint for the streaming client."""
    config = Config.from_args(args)

    service_account_info = json.load(open(config.google_credential))
    audience = "https://pubsub.googleapis.com/google.pubsub.v1.Publisher"
    credentials = jwt.Credentials.from_service_account_info(
        service_account_info, audience=audience
    )
    publisher = pubsub.PublisherClient(credentials=credentials)
    topic_name = f"projects/{config.google_project_id}/topics/{config.topic_name}"

    def publish_payload(payload: Payload) -> None:
        data = json.dumps(payload)
        _log.info("Publishing message: '%s'", data)
        future = publisher.publish(topic_name, data.encode("utf-8"))
        future.result()

    poll_av_and_publish(
        poll_interval=config.poll_interval,
        publish_payload=publish_payload,
        source_name=config.source_name,
    )


# Time for an interval of blinking, seconds
_BLINK_DURATION = 0.1
# Number of alerting blinks before solid color
_BLINK_REPEAT = 3

Rgb = Tuple[int, int, int]
_RGB_VIDEO = (255, 0, 0)
_RGB_AUDIO = (0, 0, 255)
_RGB_OFF = (0, 0, 0)


_HARDWARE_KEYS = ("audio", "video")


@dataclass(eq=True, frozen=True)
class ComputedState:
    audio: bool
    video: bool

    @staticmethod
    def from_source_states(source_states: Iterable[Payload]) -> "ComputedState":
        audio = False
        video = False

        for state in source_states:
            if state["audio"]:
                audio = True
            if state["video"]:
                video = True

        return ComputedState(audio=audio, video=video)


class DisplayState:
    _device: Optional[Blink1]
    # Map of {source name -> Payload}
    _source_states: DefaultDict[str, Payload]
    # Computed hash over all source states
    _state: ComputedState
    # Last color to display
    _last_color: Rgb

    def __init__(self, device):
        self._device = device
        self._source_states = defaultdict(dict)
        self._state = ComputedState(audio=False, video=False)
        self._last_color = _RGB_OFF

    def __enter__(self, *args):
        return self

    def __exit__(self, *args):
        if self._device:
            self._device.off()

    def _solid(self, color: Rgb) -> None:
        _log.debug("Color set: '%s'", color)
        if self._device:
            self._device.fade_to_rgb(0, *color)

    def _blink(self, color: Rgb) -> None:
        """Blink the given new color, alternating back to the current color."""
        for _ in range(0, _BLINK_REPEAT):
            self._solid(color)
            time.sleep(_BLINK_DURATION)
            self._solid(self._last_color)
            time.sleep(_BLINK_DURATION)

    def update(self, payload: Payload) -> None:
        source_name = payload["source"]
        previous_source_state = self._source_states[source_name]
        if previous_source_state == payload:
            _log.debug("Source state is unchanged: '%s'", payload)
            return
        self._source_states[source_name] = payload

        state = ComputedState.from_source_states(self._source_states.values())
        if self._state == state:
            _log.debug("Computed state is unchanged: '%s'", state)
            return
        self._state = state
        _log.debug("Computed state updated: '%s'", state)

        if state.video:
            color = _RGB_VIDEO
        elif state.audio:
            color = _RGB_AUDIO
        else:
            color = _RGB_OFF

        self._blink(color)
        self._solid(color)
        self._last_color = color


_LISTEN_SKEW = timedelta(minutes=1)


def run_listen(args: argparse.Namespace) -> None:
    """Entrypoint for the listening client."""
    service_account_info = json.load(open(args.google_credential))
    audience = "https://pubsub.googleapis.com/google.pubsub.v1.Subscriber"
    credentials = jwt.Credentials.from_service_account_info(
        service_account_info, audience=audience
    )
    subscriber = pubsub.SubscriberClient(credentials=credentials)
    subscription_name = (
        f"projects/{args.google_project_id}/subscriptions/{args.subscription_name}"
    )

    if args.no_blink1:
        device = None
    else:
        device = Blink1()
        device.off()

    with DisplayState(device) as display_state:

        def recieve_message(message) -> None:
            now = datetime.now(tz=timezone.utc)
            if message.publish_time < (now - _LISTEN_SKEW):
                _log.debug(
                    "Discarding skewed message: now '%s', message '%s'",
                    now,
                    message.publish_time,
                )
                message.ack()
                return

            payload = message.data.decode("utf-8")
            _log.info("Recieved message: %s", payload)
            data = json.loads(payload)
            display_state.update(data)

            message.ack()

        # This will loop forever, so set a nice shutdown handler
        signal.signal(signal.SIGINT, shutdown)

        _log.info("Listening for published updates")
        future = subscriber.subscribe(subscription_name, recieve_message)
        future.result()


def shutdown(signal_number, frame):
    """Shutdown handler for system signals."""
    _log.info(
        "Received signal '%s', shutting down",
        signal_number,
    )
    sys.exit(1)


def main() -> None:
    """Main command line entrypoint."""
    root_logger = logging.getLogger()
    root_logger.setLevel(logging.INFO)
    stream_handler = logging.StreamHandler()
    stream_handler.setFormatter(
        logging.Formatter("%(asctime)s|%(name)s|%(levelname)s|%(message)s")
    )
    root_logger.handlers = [stream_handler]

    parser = argparse.ArgumentParser(description="on-air warning light")
    parser.add_argument(
        "--google-project-id",
        type=str,
        help="Google project id to use",
    )
    parser.add_argument(
        "--google-credential",
        type=str,
        help="Google credential JWT file",
    )

    subparsers = parser.add_subparsers(required=True, dest="subcommand")
    stream = subparsers.add_parser("stream")
    stream.set_defaults(execute=run_stream)
    parser.add_argument(
        "--config",
        type=str,
        help="Config file to use. Does not load a config file by default",
    )
    stream.add_argument(
        "--poll-interval",
        type=int,
        help="Interval to check local state in seconds",
    )
    stream.add_argument(
        "--topic-name",
        type=str,
        help="Topic to publish messages to",
    )
    stream.add_argument(
        "--source-name",
        type=str,
        help="Specify the name of this source of events. Defaults to system hostname.",
    )

    listen = subparsers.add_parser("listen")
    listen.set_defaults(execute=run_listen)
    listen.add_argument(
        "--subscription-name",
        type=str,
        help="Google subscription to recieve messages from",
    )
    listen.add_argument(
        "--no-blink1",
        action="store_true",
        help="Do not use blink(1) for user notifications",
    )

    args = parser.parse_args()
    args.execute(args)


if __name__ == "__main__":
    main()
